### ISSUE TITLE
[Awaiting Payment 29th May] Threads - There is no Leave thread option on group Chat threads

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#39660]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.60-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** dave0123seife+hl5@gmail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
1. Go to any Group chat/ Create a group chat
2. Send a message
3. Go to threads on the sent message
4. Click on the three dots on the header, Notice that there is no Leave option.
5. Send a message on the Thread.
6. Again Click on the three dots on the header, Notice that there is no Leave option.

## Expected Result:
There should be a leave thread option after creating a thread.

## Actual Result:
There is no Leave thread option.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/4728b4f0-0d75-4d6f-962e-2e88d327d434

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by izarutskaya** [comment]
We think this issue might be related to the #vip-vsb.

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There is no Leave thread option.

### What is the root cause of that problem?
We don't allow leave or join for group chat by this check here and the `chatType` of thread is inherited by the parent so the thread of group chat doesn't have `leave` option.

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/HeaderView.tsx#L145

And I also see that the `canLeaveRoom` function return `true` for group chat because we don't check the `chatType` is `group` here
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/libs/ReportUtils.ts#L4884-L4889
### What changes do you think we should make in order to solve the problem?
1. We should remove the check `isGroupChat` here 

```
report?.chatType === CONST.REPORT.CHAT_TYPE.GROUP ||
```
https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/HeaderView.tsx#L145

2. Add the check `chatType` is group here so for the group chat, the join/leave option will not appear 

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/libs/ReportUtils.ts#L4884-L4889

and for the thread of this, join/leave option will appear by `isChatThread` condition

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/HeaderView.tsx#L145

### What alternative solutions did you explore? (Optional)
NA

**Comment by trjExpensify** [comment]
Moving this forward, and putting it in #vip-split as a groups bug. CC: @marcaaron @JmillsExpensify if you want to track it somewhere or it's already known as part of the Groups project rollout.

**Comment by trjExpensify** [comment]
Ah whoops, needs an external label innit! @jjcoffee proposals here for you to review. Thanks! :)

**Comment by jjcoffee** [comment]
Tough decision here :smile: @Nodebrute's [proposal](https://github.com/Expensify/App/issues/39660#issuecomment-2038337679) does work, but I feel that it makes the code a bit unreadable (in an already pretty hard to parse set of conditions). 

I think removing the `!isGroupChat` condition as in the other two proposals is the way to go here, as using that is what actually blocks any threaded messages from having the join/leave option (I imagine this might be on purpose for the self-DM). In that case, @nkdengineer's [proposal](https://github.com/Expensify/App/issues/39660#issuecomment-2039144454) is the one to go with as it also deals with keeping the join/leave option from appearing on the group chat itself.

:ribbon::eyes::ribbon: C+ reviewed

**Comment by iwiznia** [comment]
> We should remove the check isGroupChat here
> Add the check chatType is group here so for the group chat, the join/leave option will not appear

Isn't that basically the same thing except the new code will not check for `|| ReportUtils.isDeprecatedGroupDM(report);`?

> and for the thread of this, join/leave option will appear by isChatThread condition

I don't understand what you mean here, can you clarify? Is the only change the one above?

**Comment by nkdengineer** [comment]
> Isn't that basically the same thing except the new code will not check for || ReportUtils.isDeprecatedGroupDM(report);?


@iwiznia Yeah we should bring all checks for `isGroupChat` into `canLeaveRoom` function instead of only adding the check chatType is group

```
const isGroupChat = ReportUtils.isGroupChat(report) || ReportUtils.isDeprecatedGroupDM(report);
```

> I don't understand what you mean here, can you clarify? Is the only change the one above?

Yes, I mean after we move the check `isGroupChat` into `canLeaveRoom` function, the join/leave option will be displayed correctly for both group chat and the thread of its. 

1. The join/leave option will not appear in group chat by adding the check in `canLeaveRoom` function.


2. And the thread of group chat will appear join/leave option with the new check.

```
const canJoinOrLeave = !isSelfDM  && (isChatThread || isUserCreatedPolicyRoom || canLeaveRoom); 
```

**Comment by iwiznia** [comment]
> The join/leave option will not appear in group chat by adding the check in canLeaveRoom function.

Why would you not be able to leave a group chat? 😕

**Comment by nkdengineer** [comment]
> Why would you not be able to leave a group chat? 😕

> There should be a leave thread option after creating a thread.

@iwiznia I see that the expected here is we only want to have leave thread option in the thread of group chat.

**Comment by iwiznia** [comment]
Where are you seeing that? I think it is wrong, yes.

**Comment by nkdengineer** [comment]
@iwiznia I follow the expected of this issue.

<img width="834" alt="Screenshot 2024-04-12 at 22 54 15" src="https://github.com/Expensify/App/assets/161821005/82f8b071-8040-454a-a491-54048316cbdc">

**Comment by iwiznia** [comment]
Yes, that is correct, it is saying you should be able to leave a thread, it is NOT saying you should not be able to leave a group.

**Comment by iwiznia** [comment]
@jjcoffee did you also misinterpret the issue?

**Comment by nkdengineer** [comment]
> Click on the three dots on the header, Notice that there is no Leave option.

@iwiznia 

I see that in the step 4 notice that there is no leave option and we don't consider this as bug in Actual result. So I think the expected for group chat is no leave/join option.


I updated my alternative solution https://github.com/Expensify/App/issues/39660#issuecomment-2039144454 to allow leave/join option for both group chat and the group chat's thread

**Comment by marcaaron** [comment]
@iwiznia I can take this one if you want. 

Otherwise, the issue is about **threads in Group Chats**. The way to "leave" the Group Chat itself is via the report details page and a new "Leave" button (which will get merged soon). I believe that we will want to streamline all of this so that threads, rooms, Group Chats are all using a similar UI cc @shawnborton 

For now - users do need the ability to leave the Group Chat thread, but leaving the Group Chat the same way as the thread (three button menu) will require more of a change and I'm not sure is what we want anyway.

**Comment by iwiznia** [comment]
Sure, go ahead

**Comment by marcaaron** [comment]
@nkdengineer we are adding the ability to leave a Group Chat in [this PR](https://github.com/Expensify/App/pull/39757) (previously there was no way to leave a Group DM).

For now, I think we will want to give users the ability to leave the Group Chat threads. 

In this case, the problem is that `isGroupChat()` returns true for group chat _threads_ so the `!isGroupChat` condition overrides the `isChatThread` logic. I believe what we ultimately want is something like this:

```js
function isGroupChatThread(report: OnyxEntry<Report> | Partial<Report>): boolean {
    return isChatThread(report) && getChatType(report) === CONST.REPORT.CHAT_TYPE.GROUP;
}

function isGroupChat(report: OnyxEntry<Report> | Partial<Report>): boolean {
    return !isChatThread(report) && getChatType(report) === CONST.REPORT.CHAT_TYPE.GROUP;
}
```

However, there might be other changes to watch out for if we do this since the `isGroupChat()` method is called in several places.

**Comment by nkdengineer** [comment]
@marcaaron What do you think about my alternative solution here.

**Comment by jjcoffee** [comment]
@marcaaron Oooh thanks for the heads up on that PR! I like that idea in general of adding those extra helper functions, but I wonder if it's worth it if we will only (likely) use `isGroupChatThread` for this issue? It looks like there are cases where we currently use `isGroupChat` for either the main chat _or_ the thread, so maybe we can get away with just adding a single extra `isMainGroupChat` instead (or maybe `isParentGroupChat` is a bit better)?

```
function isMainGroupChat(report: OnyxEntry<Report> | Partial<Report>): boolean {
    return !isChatThread(report) && getChatType(report) === CONST.REPORT.CHAT_TYPE.GROUP;
}
```

**Comment by jjcoffee** [comment]
> What do you think about my alternative solution here.

@nkdengineer Your alternative solution would allow leaving the group chat through a different method from what has been added in [the PR](https://github.com/Expensify/App/pull/39757) that @marcaaron linked, so I don't think we can go with that (for now).

**Comment by marcaaron** [comment]
I like `isParentGroupChat()`. I think it could still be possible that we run into some situations where we ask if something is a "Group Chat" and it's actually a "Group Chat thread" i.e. it should behave more like a "thread" than a Group Chat. But I don't know of any specific places in the code where this is happening or where we need it.

**Comment by marcaaron** [comment]
All of this looks due for a clean up of some kind. Feels like we could also maybe just have a `canLeaveReport()` method instead of a `canLeaveRoom()` and stop doing this confusing stuff:

https://github.com/Expensify/App/blob/a7820d2027138a1a9670b204f608d6b92ec0a454/src/pages/home/HeaderView.tsx#L146-L148

I'm unsure why we are combining the concepts of "leave" and "join" they are different things entirely and I find it quite strange that `canJoinOrLeave` is declared _before_ a variable called `canLeave` and `canJoin` and we are not just setting the value of `canJoinOrLeave` to `canLeave || canJoin`. 

Maybe we can just set each one individually and have something like `canLeaveReport()` and `canJoinReport()` helper methods.

**Comment by Nodebrute** [comment]
Hey @marcaaron, I believe all types of threads should be treated equally. Why are we setting the chatType as 'group' for threads created from a group chat? Do they have any differences from threads that are not created from groups?

**Comment by Nodebrute** [comment]
@jjcoffee @marcaaron We can resolve this issue with minimal changes. Instead of creating a new helper function or altering 'isGroupChat()', what if we modify the code
https://github.com/Expensify/App/blob/90412ec0c38cc1ccd0a1198cf1f4712474667906/src/pages/home/HeaderView.tsx#L77
to 
```
const isGroupChat = !ReportUtils.isChatThread(report) && (ReportUtils.isGroupChat(report) || ReportUtils.isDeprecatedGroupDM(report));
```
With this adjustment, we can display 'leave' and 'join' options for group threads. Group threads will function like any other thread. After implementing this change, we can still leave the group from the details page.

**Comment by nkdengineer** [comment]
> @nkdengineer Your alternative solution would allow leaving the group chat through a different method from what has been added in https://github.com/Expensify/App/pull/39757 that @marcaaron linked, so I don't think we can go with that (for now).

@jjcoffee We can add the check in the callback of leave option here. We will call `leaveGroupChat` if the report is group chat report and not is a thread. Otherwise, we will call `leaveRoom`.

https://github.com/Expensify/App/blob/eef5fde8536c78726dcbcf1632839b803b7d8526/src/pages/home/HeaderView.tsx#L160

**Comment by nkdengineer** [comment]
@marcaaron 

I think `canJoinOrLeave` is declared before `canJoin` and `canLeave` because we use this condition in both `canJoin`, `canLeave`.

>  const canJoinOrLeave = !isSelfDM && !isGroupChat && (isChatThread || isUserCreatedPolicyRoom || canLeaveRoom || canLeavePolicyExpenseChat); 

if we want to remove `canJoinOrLeave` we can update like this 

```
const canLeave = !isSelfDM && ((isChatThread && !!report.notificationPreference?.length) || isUserCreatedPolicyRoom || canLeaveRoom || canLeavePolicyExpenseChat);
const canJoin = canLeave && !isWhisperAction && report.notificationPreference === CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN && (!isGroupChat || isChatThread);
```

**Comment by marcaaron** [comment]
I think we have enough context to start working on this now. @nkdengineer can you summarize what your final approach will be please to make sure we are on the same page before raising a PR?

**Comment by nkdengineer** [comment]
@marcaaron Here is my final approach. Let me know if anything is still confusing


1. Update here like this to allow the join/leave option appears in group chat's thread and leave option appear in group chat.
```
const canLeave = !isSelfDM && ((isChatThread && !!report.notificationPreference?.length) || isUserCreatedPolicyRoom || canLeaveRoom || canLeavePolicyExpenseChat);
const canJoin = canLeave && !isWhisperAction && report.notificationPreference === CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN && (!isGroupChat || isChatThread);
```

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/HeaderView.tsx#L145-L147

2. Update the `onSelected` function here to call `leaveGroupChat` if the report is group chat otherwise call `leaveRoom` function.

```
onSelected: Session.checkIfActionIsAllowed(() => {
    if (isGroupChat && !isChatThread) {
        Report.leaveGroupChat(reportID);
        return;
    }
    Report.leaveRoom(reportID, isWorkspaceMemberLeavingWorkspaceRoom)
}),
```

https://github.com/Expensify/App/blob/f95bf85925686e64bf6c5bcfb6e66722ff6934fc/src/pages/home/HeaderView.tsx#L159

**Comment by marcaaron** [comment]
Sorry I don't think you are understanding the assignment entirely. To clarify, the goal of this ticket is to include the "leave" option for threads inside Group Chats. Not add it for Group Chats themselves. So, part 2 of your proposal is incorrect. We should never be reaching that code when we have a Group Chat. Thanks!

**Comment by nkdengineer** [comment]
> To clarify, the goal of this ticket is to include the "leave" option for threads inside Group Chats. Not add it for Group Chats themselves

@marcaaron Got it now. With this expected behavior and your suggestion here https://github.com/Expensify/App/issues/39660#issuecomment-2058029341. We can do with this approach:

1. Create `isParentGroupChat` util

```
function isParentGroupChat(report: OnyxEntry<Report> | Partial<Report>): boolean {
    return !isChatThread(report) && (isGroupChat(report) || isDeprecatedGroupDM(report));
}
```

2. Update `canJoinOrLeave` to 

```
const canJoinOrLeave = !isSelfDM && !isParentGroupChat(report) && (isChatThread || isUserCreatedPolicyRoom || canLeaveRoom); 

```

or remove `canJoinOrLeave` and update like this.

```
const canLeave = !isSelfDM && !isParentGroupChat(report) && ((isChatThread && !!report.notificationPreference?.length) || isUserCreatedPolicyRoom || canLeaveRoom || canLeavePolicyExpenseChat);
const canJoin = canLeave && !isWhisperAction && report.notificationPreference === CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN && (!isGroupChat || isChatThread);
```

**Comment by Nodebrute** [comment]
@marcaaron @jjcoffee Hey, could you take a look at this [comment](https://github.com/Expensify/App/issues/39660#issuecomment-2058275743)? With this approach, we don't need to create a separate utility function. We're just making changes within this file, and it seems to be working fine.

**Comment by marcaaron** [comment]
@nkdengineer Thanks, looks good, let's do it. I'll give more notes in the PR review.

**Comment by jjcoffee** [comment]
> With this approach, we don't need to create a separate utility function.

@Nodebrute I don't think there's any reason to avoid creating utility functions, we actively want them where they're useful. In this case having `isParentGroupChat` instead of a modified `isGroupChat` (which otherwise means different things in different contexts) is much clearer, so it's beneficial to have it.

**Comment by nkdengineer** [comment]
I'm working on the PR, will open this early in the morning.

**Comment by nkdengineer** [comment]
@jjcoffee The PR is [here](https://github.com/Expensify/App/pull/40529).

**Comment by trjExpensify** [comment]
PR was merged earlier today, Melv.

**Comment by saracouto** [comment]
@trjExpensify what is left to be done here?

**Comment by trjExpensify** [comment]
The PR hit production 2 days ago, so we'll now wait 5 more days for any regressions. Then we'll pay it out and close. :)

**Comment by jjcoffee** [comment]
- The PR that introduced the bug has been identified. Link to the PR: N/A - this has always been the behaviour as far as I can tell
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- Determine if we should create a regression test for this bug. **Yes**
- If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression Test Proposal**

1. Create a group chat
2. Go to the group chat and send a message
3. Click on three-dot menu in the header
4. Verify that leave option doesn't appear
5. Reply in thread to the group chat message in (2)
6. Repeat step 3 and verify that the leave option appears

Do we agree 👍 or 👎

**Comment by marcaaron** [comment]
LGTM @jjcoffee

**Comment by trjExpensify** [comment]
Thanks for [the steps](https://github.com/Expensify/App/issues/39660#issuecomment-2135364864)! Why can't you leave a group in step #3 though? I thought anyone could leave those now. 🤔 

If that's accurate, my suggested edit would be: 

1. Create a group chat
2. Send a message in the group chat
3. Click on the three-dot overflow menu in the header of the group chat
4. **Verify** the leave option appears
5. Reply in thread to the group chat message sent in (2)
6. Click on the three-dot overflow menu in the header of the thread
7.  **Verify** that the leave option appears

**Comment by jjcoffee** [comment]
@trjExpensify Ah yeah you _are_ right that you can leave groups but it's not via the three-dot menu. You have to tap the header and then the modal has the leave option:

![image](https://github.com/Expensify/App/assets/27287420/92482e01-3958-4f78-85b8-0270b394c35e)

We could add an extra step to tap the header and verify that the leave option is there? It'd be good to keep the step verifying that the leave option isn't in the three-dot menu though.

**Comment by trjExpensify** [comment]
Hm, so actually.. we're in the process of updating all of these in a project called "Details revamp" in implementation. So there won't be a overflow menu in any of the headers, the buttons will reside on the `Details` page instead. 

So I'm going to hold off on this regression test, and the project when wrapped will add/update them all centrally.

**Comment by trjExpensify** [comment]
Payment summary as follows: 

- $500 to @nkdengineer for the fix
- $500 to @jjcoffee for the C+ review 

Offers sent!

**Comment by trjExpensify** [comment]
Joel, paid!

**Comment by nkdengineer** [comment]
@trjExpensify Thank you, I accepted the offer

**Comment by trjExpensify** [comment]
Perfect, paid. Closing!
